runtime/service_integ_test.go

```diff
@@ -77,6 +78,40 @@ func TestShimExitsUponContainerKill_Isolated(t *testing.T) {
 	image, err := client.Pull(pullCtx, debianDockerImage, containerd.WithPullUnpack, containerd.WithPullSnapshotter(naiveSnapshotterName))
 	require.NoError(t, err, "failed to pull image %s, is the the %s snapshotter running?", debianDockerImage, naiveSnapshotterName)
 
+	rootfsBytes, err := ioutil.ReadFile(defaultVMRootfsPath)
+	require.NoError(t, err, "failed to read rootfs file")
+
+	const vmID = 0
+
+	tapName := fmt.Sprintf("tap%d", vmID)
+	err = createTapDevice(ctx, tapName)
+	require.NoError(t, err, "failed to create tap device for vm %d", vmID)
+
+	// TODO once Noah's immutable rootfs change is merged, we can use that as our rootfs for all VMs instead of copying
+	// one per-VM
+	rootfsPath := fmt.Sprintf("%s.%d", defaultVMRootfsPath, vmID)
+	err = ioutil.WriteFile(rootfsPath, rootfsBytes, 0600)
+	require.NoError(t, err, "failed to copy vm rootfs to %s", rootfsPath)
+
+	fcClient := fccontrol.NewFirecrackerClient(client.Conn())
```

Comment:
Why is the TestShimExitsUponContainerKill_Isolated test being updated with a CreateVM call? I'd prefer we keep it as it was previously as it gives us coverage of the codepath where a VM gets created implicitly as part of CreateTask

Reply:
This is due to container count being 0. Saw your comment above regarding setting the `ContainerCount`. However, I don't know if the logic is right. Basically we need to get the container count value somewhere, currently I have it set to zero unless `vmID` is `""` which then sets `ContainerCount` to `1`. However, my impression was we'd have more than one container in a VM, so why are we only setting it to 0 or 1?

Reply:
What I was imagining is this:
1. If no vmID was passed, we can just set it to 1 since there will only be one task in the implicitly created VM.
2. If a vmID was passed and the VM was pre-created with a `CreateVM` call, then ShimStart just gets an `AlreadyExists` error, so it doesn't matter what value the ContainerCount is (may as well be 0).
3. If a vmID was passed but the VM was NOT pre-created, then the VM should just be created with default options (default rootfs, kernel, other settings, etc). ShimStart can thus leave the ContainerCount value as 0 in this case and the control plugin should handle setting the default.

I think given the above, the only extra required change on top of what you already did would be having the control plugin set a default ContainerCount if it receives a 0 value. This could be either a hardcoded default or maybe one configurable through the runtime config file.

I'm not sure what would be a good default value. Isn't there a maximum number of possible drives currently that's relatively low? Maybe that's a good starting place for now? Don't have a strong opinion on this yet though.

Reply:
@sipsma - I think I'll keep it at `1` for now until we have good reason to change it. Any objections?

Reply:
Yeah, it's an obscure case so as long as there's a clear error when a user tries to spawn another container in the VM that sounds good to me as a starting point.

Reply:
@xibz Do we have a test for the error condition?

Reply:
@samuelkarp - we do now :D

Reply:
I believe we still need an update to the code to handle case 3 I listed above, right now in that case the container count will stay 0 incorrectly. That's why I was suggesting we set a default value if the control plugin receives a value of 0 for ContainerCount. That handles not only case 3 but also the case where a user fails to provide a ContainerCount value in a direct CreateVM request.

Reply:
@sipsma - Fixed

Reply:
The fix works but I still prefer it be done by setting a default in the case where 0 is received in the request (either the control plugin or, perhaps even better, the place where we set all other default VM configuration values in `buildVMConfiguration` around line 618 of `runtime/service.go`).

That handles not only case 3 but also the case where a user fails to provide a ContainerCount value in a direct CreateVM request.

Reply:
@sipsma - Fixed. However, I generally don't like 0 being considered as unset, but since proto3 requires it as such, not much we can do.

Reply:
Setting the default in `buildVMConfiguration` looks good but we shouldn't remove what you originally had in `StartShim` where it's also explicitly set to 1 for the `vmID == ""` case.

That `vmID == ""` case should always absolutely be 1, however the default value being set in `buildVMConfiguration` could easily change in the future from 1 to something else (it's more or less arbitrary right now), at which time we'd need to remember to go back and change StartShim again. The fact that both places are setting 1 right now is more or less a coincidence. It'd be best to just explicitly set it in both places IMO.